chia/full_node/mempool_manager.py

```diff
@@ -142,34 +143,61 @@ async def create_bundle_from_mempool(
 
         cost_sum = 0  # Checks that total cost does not exceed block maximum
         fee_sum = 0  # Checks that total fees don't exceed 64 bits
-        spend_bundles: List[SpendBundle] = []
+        processed_spend_bundles = 0
         removals = []
         additions = []
+        dedup_coin_spends: Dict[bytes32, uint64] = {}
+        agg = SpendBundle([], G2Element())
         broke_from_inner_loop = False
         log.info(f"Starting to make block, max cost: {self.constants.MAX_BLOCK_COST_CLVM}")
         for dic in reversed(self.mempool.sorted_spends.values()):
             if broke_from_inner_loop:
                 break
             for item in dic.values():
-                log.info(f"Cumulative cost: {cost_sum}, fee per cost: {item.fee / item.cost}")
+                item_cost = item.cost
+                dedup_spends = False
+                # See if this item has a coin spend that is eligible for deduplication
+                potential_coin_spends = [s for s in item.npc_result.conds.spends if s.eligible_for_dedup]
+                if len(potential_coin_spends) > 0:
+                    # Current iteration only deals with a bundle with only one spend eligible for deduplication
+                    coin_spend_in_bundle = next(
+                        cs for cs in item.spend_bundle.coin_spends if cs.coin.name() == potential_coin_spends[0].coin_id
+                    )
+                    coin_spend_solution_hash = coin_spend_in_bundle.solution.get_tree_hash()
+                    # See if we processed an item with this coin spend before
+                    if coin_spend_solution_hash in dedup_coin_spends:
+                        # Process this item with this reduced cost in mind
+                        duplicate_cost = dedup_coin_spends[coin_spend_solution_hash]
+                        item_cost -= duplicate_cost
+                        dedup_spends = True
+                    else:
+                        # Process this item normally but store the isolated spend cost future reference
+                        # TODO: Ask Arvid whether INFINITE_COST is adequate here (and whether this is how to get the cost in the first place)
+                        coin_spend_cost = uint64(
+                            coin_spend_in_bundle.puzzle_reveal.run_with_cost(
+                                INFINITE_COST, coin_spend_in_bundle.solution
```

Comment:
the cost limit here should be the cost of the whole spend-bundle. Just running one solution should never exceed that. Technically, a solution shouldn't make it in here if it's too costly either, but using `INFINITE` makes me nervous. We should set it to the true upper limit of this solution.